src/treelearner/serial_tree_learner.h

```diff
@@ -180,6 +189,13 @@ class SerialTreeLearner: public TreeLearner {
   std::vector<int> ordered_bin_indices_;
   bool is_constant_hessian_;
   std::unique_ptr<CostEfficientGradientBoosting> cegb_;
+
+  std::vector<std::vector<double> > dummy_min_constraints;
+  std::vector<std::vector<double> > min_constraints;
+  std::vector<std::vector<double> > dummy_max_constraints;
+  std::vector<std::vector<double> > max_constraints;
+
+  CurrentConstraints current_constraints;
```

Comment:
`current_constraints` may need to sync among machines in distributed learning, refer to `min_constraint` and `max_constraint` in SplitInfo.

Reply:
Could be more specific about distributed learning? Our method already supports having several threads, because leaves are treated separately, and each leaf gets associated to 1 array (we use "int tid" to identify what thread we are in).

Reply:
I mean the multi-machine, refer to the states you delete in SplitInfo. The SplitInfo will be sync in different nodes.

Reply:
I am sorry but I don't understand. My understanding is that the object current_constraints contains splitting_constraints_vector which is an array which contains as many elements as there are threads. Each element of the array is used in one specific thread, and does not need to communicate with the others because the computations for 1 leaf is made by at most 1 thread. 

Can you be more specific about the part of SplitInfo that is relevant? Is it relevant with all tree learners, or only with some? I have only been using serial_tree_learner, so I don't really know a lot about the others. Thanks

Reply:
Refer to the DataParallelTreeLearner or other parallel learners, these learn will find the best split in different feature sets, over different machines. And these states in different machines cannot be accessed directly. We need to sync them explicitly. And SplitInfo will be synced in these parallel learners. So the min_constraints, and other states you delete in SplitInfo, is not synced anymore..

Reply:
Okay, so if I am following correctly, I need to make a functions like SyncUpGlobalBestSplit, CopyFrom, CopyTo (from splitInfo), but for my variables constraints_per_leaf_ which is an array of LeafConstraints is that right? This variable contains the constraints for all leaves in the tree, and I think it is the only one that needs to be shared. Do you agree?

Reply:
yes, the easiest way is to save the constraints into SplitInfo, then you don't need to care these. The splitInfo is per leaf too, so you can only save the information of current leaf to it, and get it back to your `constraints_per_leaf_`.

Reply:
I tried to look a bit at that part of the code but I am very unfamiliar with it. Could you explain to me how the multi-machine works overall? For example what tasks are split among machines, why we need to sync up SplitInfo, but not other variables... Also, it seems like only 2 SplitInfo are being sync'ed across machines (smaller_best_split and larger_best_split). However, in our constraining methods, a split on a leaf can generate an update in the constraints of many other leaves, so I think I still need to sync all constraints. Overall this task seems pretty difficult for me, do you think you could provide help for the actual coding? Thanks!

Reply:
I see. Maybe we can ignore this part first. Let us focus on other parts first.

Reply:
Okay, I think I pushed changes to answer about all your other comments, let me know what else I can improve!